### PR TITLE
Enable ARM64 Workflow

### LINKED ISSUE #122: Enable ARM64 GitHub workflows
This is nice to have, as discussed in the fortnightly etcd community meeting. But given that etcd has official support for ARM, it's important to run GitHub workflows using ARM64 runners.

This would be a good candidate for the good first issue label.

Refer to etcd-io/etcd#16801, etcd-io/bbolt#583 and the Pull Requests that close those issues.

### PR DESCRIPTION
**Pull request by tcchawla**
Enables test runners to use ARM64 in the workflow. Fixes #122

### CODE REVIEW

**Comment by tcchawla** [comment]
Hey @jmhbnz, I have made the necessary changes, please approve the workflow to execute the pending jobs.

File: .github/workflows/test_arm64.yaml
Comment (jmhbnz) [inline .github/workflows/test_arm64.yaml:13]: Based on the existing tests in https://github.com/etcd-io/raft/blob/main/.github/workflows/test.yaml#L10-L11 I think we only need to enable `linux-arm64-unit-4-cpu-race`

File: .github/workflows/test_arm64.yaml
Comment (jmhbnz) [inline .github/workflows/test_arm64.yaml:5]: Suggest we reduce this to `actuated-arm64-8cpu-8gb`

File: .github/workflows/test_arm64.yaml
Comment (jmhbnz) [inline .github/workflows/test_arm64.yaml:18]: Let's update to `1.21.6` which has just come out.

File: .github/workflows/test_arm64.yaml
Comment (jmhbnz) [inline .github/workflows/test_arm64.yaml:32]: As above - suggest we simplify this to just do the `4-cpu-race` target.

**Review by jmhbnz — COMMENTED** [review]
Thanks @tcchawla a few points of feedback below.

**Comment by tcchawla** [comment]
Hey @jmhbnz, I have addressed these review comments in the latest commit. Please verify.

**Comment by ivanvc** [comment]
@jmhbnz, would it make sense to introduce a template so ARM64/AMD64+i386 can reuse the same base and keep the workflow DRY?

**Comment by jmhbnz** [comment]
> @jmhbnz, would it make sense to introduce a template so ARM64/AMD64+i386 can reuse the same base and keep the workflow DRY?

That's a great suggestion, happy for it to be done in this pr or alternatively as a follow-up.

**Comment by tcchawla** [comment]
I can do it in this PR if that's okay

**Comment by tcchawla** [comment]
Hey @jmhbnz, I have added a template file and made changes to the workflows test files. Let me know if any changes are required.

**Review by jmhbnz — COMMENTED** [review]
Hey @tcchawla - If you take a look at the actions run you will see some errors: https://github.com/etcd-io/raft/actions/runs/7586269735 

```bash
 Invalid workflow file: .github/workflows/test_arm64.yaml#L6
The workflow is not valid. In .github/workflows/test_arm64.yaml (Line: 6, Col: 11): Error from called workflow etcd-io/raft/.github/workflows/test_template.yaml@11d03fd542a2cc6c8cd869e5a45b1634ef9261a1 (Line: 21, Col: 17): Unrecognized named-value: 'input'. Located at position 10 within expression: fromJSON(input.targets) In .github/workflows/test_arm64.yaml (Line: 6, Col: 11): Error from called workflow etcd-io/raft/.github/workflows/test_template.yaml@11d03fd542a2cc6c8cd869e5a45b1634ef9261a1 (Line: 21, Col: 17): Unexpected value '${{ fromJSON(input.targets) }}'
```

Looks like the templating is not quite right for the targets value.  Please give it another attempt and feel free to message me or @ivanvc on kubernetes slack if you get stuck.

**Comment by tcchawla** [comment]
Hey @jmhbnz, turns out it was a typo. I have pushed the working code.

File: .github/workflows/test_template.yaml
Comment (jmhbnz) [inline .github/workflows/test_template.yaml:30]: Looks like the `arm64` test is showing unexpected caching behaviour: https://github.com/etcd-io/raft/actions/runs/7589241324/job/20673470700?pr=128 

> ```bash
> Starting at: Fri Jan 19 21:14:04 UTC 2024
> 'unit' started at Fri Jan 19 21:14:04 UTC 2024
> ok  	go.etcd.io/raft/v3	2.651s
> ok  	go.etcd.io/raft/v3/confchange	1.468s
> ok  	go.etcd.io/raft/v3/quorum	2.166s
> ok  	go.etcd.io/raft/v3/raftpb	(cached)
> ok  	go.etcd.io/raft/v3/rafttest	(cached)
> ok  	go.etcd.io/raft/v3/tracker	(cached)
> 'unit' completed at Fri Jan 19 21:14:07 UTC 2024
> SUCCESS
> ```


Let's prevent that as follows:

```suggestion
          go clean -testcache
          case "${TARGET}" in
```

File: .github/workflows/test_arm64.yaml
Comment (ivanvc) [inline .github/workflows/test_arm64.yaml:22]: I don't think we need coverage for ARM64. Running that in the AMD64 workflow should be enough. So, I suggest deleting it.

File: .github/workflows/test_template.yaml
Comment (jmhbnz) [inline .github/workflows/test_template.yaml:37]: This new target is named slightly differently, let's keep them consistent:

```suggestion
            linux-arm64-unit-4-cpu-race)
```

File: .github/workflows/test_template.yaml
Comment (ivanvc) [inline .github/workflows/test_template.yaml:12]: These default values don't match any of the targets available. I suggest either setting the default as empty (`"[]"`) or to one target that actually exists.

**Review by ivanvc — COMMENTED** [review]
Hi @tcchawla and @jmhbnz, I took a pass on this PR. Hope it's fine :v:

**Review by ivanvc — COMMENTED** [review]
Thanks for the changes. I defer to @jmhbnz and an active project maintainer :)

**Review by jmhbnz — APPROVED** [review]
LGTM - Thanks @tcchawla and @ivanvc

**Review by ahrtr — APPROVED** [review]
lgtm

Thanks all

**Pull request merged**
